### DISCUSSION TITLE
Trade-offs between ORM / Core

### DISCUSSION THREAD

**Discussion by ccrvlh** [sqlalchemy/sqlalchemy#8272] [Usage Questions] [answered]
We have some functions that are more sensitive to performance, and we are looking for alternatives to speed things up. Looking at the Core vs ORM performance comparison, I’m trying to better understand the ORM/Core paradigm, specially now that the query API is mostly unified.

As I understand, ORM is full. Fledged mapper, while Core is mostly a toolkit and query builder, but still unsure on the details.

Our most common patterns:

- 	 We usually declare our models  as a Class, the normal declarative way, is nice, but we wouldn’t mind any other way to declare the tables. So I guess we wouldn’t lose much by giving up the ORM here.
- 	 We are really enjoying the new syntax, a lot closer to SQL and more intuitive in our opinion. So we don’t miss the ORM query style at all. Again, Core would mostly be fine here.
- 	 We don’t use ORM object manipulation patterns (like `session.add` or `obj.name = “New Name; session.commit()` or the automagic relationship pattern like `obj.relationA.append(relationB)`). Actually we do prefere to manipulate the values manually. Again, Core should be ok also.
- 	 We are heavy users of `joinedload/subqueryload` , this is something that we would like to keep. This is probably where ORM shines.
- 	 We really appreciate the nested relationship dynamic (we don’t care about the class itself, but having the results organized in a nested way is great (think of a List[Address] inside the User object, but if was a nested dict would be great already). Probably also a powerful ORM feature?
- 	 We don’t use any mapped relationship pattern, we prefer to control what is being loaded when writing the query. Maybe same thing with core.

We also found that it was possible to leverage the core’s performance for really simple stuff using `MyORMTable.__table__` but it’s not very simple or intuitive for our current patterns in more complex cases. We also use a lot of `returning(Model)` during `inserts` and ORM-enabled `updates`, and we like the `namedtuple` with `._mapping` attribute.

Now I wonder what exactly would be giving up if we ditched the ORM all together? I confess it’s not very clear to me where the Core ends the ORM heavy lifting starts.

At first I though probably the nested relationships and  `joinedload/subqueryload` were more like query builder helpers and not so intricate to the ORM, but as I couldn’t find any mention to it in the Core docs, it does seems that they play a big role behind the scenes.

Any ideas or suggestions? Is my understanding correct? Are there patterns to solve those most common use cases in Core in an “easy” way using Core? Or perhaps it could be an option to mix both and use core on the most performance-critical services with `ORMModel.__table__`?

With the unified query style on 1.4-2.0 (which we really liked) I wonder what are the next steps, I mean, maybe in the future Core users can write declarative tables? And ORM users could avoid the automatic parsing to the model? Or should we expect this distinction to continue for good?

Thanks a lot.

**Comment by zzzeek** [comment, marked as answer]
I think we need to understand what the performance concerns are.     like, using ORM models up front for a mostly-Core application, totally fine and I would think this is common.  ORM models are declared at the module level and only incur overhead when the application first starts up.    so if your application is one that is long running, there's no issue here.

then, with the query unification and caching, there's similarly not much difference in executing statements with either the Connection or the Session.    The ORM's process here of converting the select() into a SQL string has a big extra step that is expensive, but it's cached now.  if you look in your logs and see "cached" for the statement, that expense has been squashed down to something not much different than straight Core.

So that leaves the only big difference is in fetching rows.  That's still more expensive with ORM if you are returning objects, and if you are into joinedload()  / selectinload(), you have to be getting objects back.     We continue to seek to improve this and the introduction of cython in 2.0 means that in 2.1 and forward we have more opportunities to convert more intricate ORM loading code into cython.   

It seems to me it would be very reasonable to use the ORM mostly and only if you have some SELECT that has lots of rows, you can use session.connection().execute(stmt) to cut out the overhead for that particular statement.      this is the idea of the select() unification, that it should be easier to move between these two usage patterns.

overall though, when someone says "performance" it should be accompanied by the word "measurement".    When things are slow, there's lots of reasons for that and lots of ways to solve the problem that are always very specific to what's going on.  I wouldnt try to say, "we only use API X and not Y because Y is slow".   that's too coarse-grained of a generalization to be practical.

**Reply by ccrvlh** [reply]
We have some services that deal with time series, not much data for the database, but can make the difference (think like 150ms vs 700ms). We normally perform some calculations on the data (user inputs) and then update rows at the same time. In our specific scenario, I think using the word "performance" isn't really the right way to put it, we don't really have a performance issue per se.

We also have been using more serialization libraries (eg Pydantic) around our code, and it happens that the ORM object itself end up not being used much at all. There's also a limitation with Pydantic when converting from SQLAlchemy to Pydantic models (Pydantic's `from_orm` does validation on the code, which means we are validating data coming from the database, an unnecessary step), sure this is a Pydantic issue, but it all comes together and it keeps adding up more milliseconds to our response time. If we want to skip the validation we still have to convert the ORMModel to a `dict` or something, which makes me wander why do want to get the Model back after all if a dict, tuple or a row would be just as useful?

I think a better way to put it is that we got the impression that we were not leveraging all of ORM's benefits, but we were paying it's full price (which is not high anyways, but it's something). So my question at the end of the day is, what exactly would I be giving up? By your answer it seems that `subqueryload/selectinload` is the most important feature I would miss by stop using the ORM, I could get away with nesting e deduplicating joined queries pretty easily I guess.

And just to make sure my understanding is correct, if I declare my models using the declarative way, the `ORMModel.__table__` object is the exact same thing as if I had declared the Table manually as a core object? Or is there any specifics I'm missing? By inspecting the element it seems so, but I haven't spent much time using Core, so I'm not totally sure.

Thanks for you time @zzzeek

**Reply by zzzeek** [reply]
> We also have been using more serialization libraries (eg Pydantic) around our code, and it happens that the ORM object itself end up not being used much at all. There's also a limitation with Pydantic when converting from SQLAlchemy to Pydantic models (Pydantic's `from_orm` does validation on the code, which means we are validating data coming from the database, an unnecessary step), sure this is a Pydantic issue, 

it is, and I hope in a few years something better has been figured out on this front.   I would suggest looking at [SQLModel](https://sqlmodel.tiangolo.com/) perhaps.

> If we want to skip the validation we still have to convert the ORMModel to a `dict` or something, which makes me wander why do want to get the Model back after all if a dict, tuple or a row would be just as useful?

sure if you're using pydantic and want to skip ORM models, get the dict back from pydantic and feed those into insert() / update() / delete() statements.  that's fine.   You're writing a little "mini-ORM" yourself but that is exactly why SQLAlchemy is broken up into Core / ORM with a pretty hard firewall between them, because this is the reality.

> I think a better way to put it is that we got the impression that we were not leveraging all of ORM's benefits, but we were paying it's full price (which is not high anyways, but it's something). 

if you have the models set up as ORM models and then you just use ``__table__``, as long as your application is not sensitive to startup time and perhaps resident memory use (if you had thousands of tables), that aspect of it doesn't really cost anything.

> So my question at the end of the day is, what exactly would I be giving up? By your answer it seems that `subqueryload/selectinload` is the most important feature I would miss by stop using the ORM, I could get away with nesting e deduplicating joined queries pretty easily I guess.

I think writing SELECT queries with the model classes vs. the plain Table objects is much better.  Especially in 2.0 you will get full pep-484 typing support (think IDE showing you all the attribute types etc. as you hover) if you query using models and not tables.  an ORM-based select() construct can still be fed into Connection.execute() where there's no ORM object-loading overhead.

> 
> And just to make sure my understanding is correct, if I declare my models using the declarative way, the `ORMModel.__table__` object is the exact same thing as if I had declared the Table manually as a core object? 

it's the same yup


I still advise doing some performance measurement when deciding on approach.   in all cases make sure whatever function/method you are timing is given a few runs to get the caches warmed up then see what you get.

**Reply by zzzeek** [reply]
oh also I highly recommend looking through the [performance examples](https://docs.sqlalchemy.org/en/14/orm/examples.html#module-examples.performance) , which will give you a complete catalog of "all the different ways to query" for all major use paradigms, that you can then run and time in different ways to see very clearly what kinds of scale are involved.

**Reply by ccrvlh** [reply]
So far I found Pydantic a very convenient and helpful data container. Apart from the validation in the application's boundaries (validating external data), we end up using a lot of helper methods for internal usage (`dict()`, `json`, `exclude_none`, `alias`, `default_factory`, etc), a few of them are present in the native dataclass, but not all. Even `from_orm` is very helpful since it parses a model tree recursively, which can be a bit of a pain to do manually depending on the data format. Haven't used much `attrs` though, I ended up liking the way pydantic leverages native typing. I think one of the biggest limitation for now is the fact the the parsing and serialization are still totally coupled with the validation, so you end up validating things a lot more than necessary. SQLModel also has this problem. I also tend to prefer to use SQLAlchemy directly rather than adding layers on top of it, but I can see how the lib will be appealing to other users, specially in small scale projects.

The new query style has been great. The autocomplete and how natural if feels compared to raw SQL has improved our experience using the lib by ten fold at least, really happy with the changes, awesome design, congrats. I didn't know I could use the ORM new query style with `connection()` to leverage core, nice to know. Ran a couple of tests with our data, and got a 3x performance gain on some queries. To be a fair comparison I would have to account to all the work I would do manually (like nesting stuff), in one hand is hard to imagine I could write a more efficient code when compared to the ORM, at the other, I would still need to handle JSON serialization anyways, so I need an extra step. If pydantic had a `skip_validation` argument to `from_orm` (today `from_orm` necessarily validates data, and the `construct` method, that doesn't validate data. is not recursive) I imagine the stack would account for a good portion of the use cases. BTW, is there specific flow you would recommend and are happy with when it comes to the cycle DB -> JSON? Maybe I'm missing something that we could do differently (and better).

I didn't know about the performance examples, (I end up discovering a new thing every day on the docs!), will most definitely dig deeper, just found out that `insert(ORMModel).values()` ends up being slower than `ORMModel.__table__.insert().values()` so definitely there are small changes to make if we were to squeeze every `ms` of performance out of our code (not the case for us at the moment).

Anyways, thanks for your time @zzzeek, I will keep experimenting with the tradeoffs.
